### PR TITLE
Add support for the kakoune text editor

### PR DESCRIPTION
**Pull request by loewenheim**
This is my first contribution to `home-manager`, so I’m not sure about some of the things I’ve done. In particular, I’m not sure about how to handle the following issue: The `tabstop` option, for instance, is set to `8` by default in `kakoune` and this value is assumed if not overriden in `kakrc`. Should I therefore:

- Make a nix option of type int with default 8 and always emit it to `kakrc`?
- Make a nix option of type int with default 8 and only emit it to `kakrc` if it differs from the default?
- Make a nix option of type null or int with default null and only emit it to `kakrc` if it isn’t null?

Right now I’ve chosen the second option, was that correct?.

### CODE REVIEW

**Comment by sboosali** [comment]
imo, #3 is best (`null`), with the default value mentioned in the
docstring. afaik, most options are "omitted" if null. and it's more future
proofed (e.g. if the underlying application changes its default). i feel
like subtle issues might come up if a user sets `8` explicitly, expecting
it to work, but then is confused when it doesn't (e.g. if there's a
global/system config that has a `7`, which the user isn't aware of, and
which #2 may run into.)

(but I haven't contributed to home-manager yet, so take this with a grain
of salt).

**Comment by rycee** [comment]
Both option 1 and 3 are OK with me. Home Manager already contain both these techniques.

I sort of am partial to 1 since it leads to nicer documentation and it is clearer for the user which value will be used in the end. But I think some others disagree with me on this. So I think as the module writer you can choose 🙂

(as long as it's not option 2)

File: modules/programs/kakoune.nix [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+  cfg = config.programs.kakoune;
+
+  configModule = types.submodule {
+    options = {
+
+      tabStop = mkOption {
+        type = types.ints.unsigned;
+        default = 8;
+        description = ''
+          The width of a tab
```
Comment (rycee) [inline modules/programs/kakoune.nix]: All option description should end in a period. I.e., here it should be `The width of a tab.`.

**Comment by loewenheim** [comment]
I’ve now changed most of the options to be in accordance with variant 3 above. It didn’t seem sensible to make `indentWithTabs` and `incrementalSearch` have type `nullOr bool`, though. Is that acceptable?

**Comment by loewenheim** [comment]
How can I check the documentation that I’ve written? I can’t find a way to compile it.

**Comment by rycee** [comment]
Sorry for the delay! The easiest way to see the documentation is to use `home-manager switch` using your repo clone. Then `man home-configuration.nix` will contain your changes. You can also set
```
manual.html.enable = true;
```
in your Home Manager configuration and after `home-manager switch` you should be able to run
```console
$ home-manager-help
```
to open the HTML-manual in a browser.

**Comment by loewenheim** [comment]
Thank you very much for the information. There were definitely some issues left with the documentation ;)

File: modules/programs/kakoune.nix [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+          <filename>~/.config/kak/kakrc</filename>.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable(mkMerge [
+    {
+      home.packages = [ pkgs.kakoune ];
+      xdg.configFile."kak/kakrc-test".source = configFile;
```
Comment (teto) [inline modules/programs/kakoune.nix]: fix the filename and we can merge since previous reviewers seem ok with the module.

**Comment by rycee** [comment]
Thanks! Rebased to master in 7d68c46feb845c572ef335f824062f90fdebf655.

Note, I made a few minor changes and reformattings. Hopefully nothing broke 🙂

**Pull request closed without merging**